### PR TITLE
[WIP] Accumulators

### PR DESCRIPTION
**Pull request by jcoughlin11**
## PR Summary

This PR adds the ability for yt to integrate any field along an arbitrary path and accumulate the results of this integration along the way.

The tests need to be finished and the code cleaned up to pass flake8, but the core functionality is present.
## PR Checklist

- [ ] Code passes flake8 checker
- [ x] New features are documented, with docstrings and narrative docs
- [ x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by matthewturk** [comment]
@jcoughlin11 can you merge with upstream yt-4.0?

File: yt/visualization/accumulators.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+    -------
+        accum : YTArray 
+            The cumulative value of the field integral at each path
+            segment
+    """
+    # https://en.wikipedia.org/wiki/Line_integral
+    # np.dot doesn't combine units correctly and there does not appear to be a
+    # unyt implementation of dot, so units will be handled manually
+    # for the time being
+    accum = np.cumsum(np.diag(np.dot(field_vals[:-1].d, (path[1:].d - path[:-1].d).T)))
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:81]: `path` is unitless, right?  That's why this works?

File: yt/visualization/accumulators.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    ..math::
+    
+        I = \int_C \phi(x1,x2,...,xn)d\vec{r}
+
+    Parameters
+    ----------
+    p : YTArray 
+        The path to be integrated along
+
+    fieldVals : YTArray
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:101]:
```suggestion
    field_vals : YTArray 
```

File: yt/visualization/accumulators.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+    # np.linalg.norm returns a ndarray, so when multiplied by field_vals, which
+    # is a YTArray, this leads to incorrect units. There does not appear to be
+    # a unyt implementation of norm, as far as I'm aware, so units will be
+    # handled manually for the time being
+    accum = np.cumsum(field_vals[:-1].d * np.linalg.norm(p[1:].d - p[:-1].d, axis=1))
+    accum = YTArray(accum, field_vals.units * p.units)
+    return accum
+
+
+def get_row_major_index(ncells, cell_ind):
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:122]: Is this the same functionality as `np.unravel_index` and `np.ravel_multi_index`?  If so, we might be able to ditch it.

File: yt/visualization/accumulators.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+    fields along various paths.
+
+    The class takes in a list of user-defined paths and a dataset. From
+    these, the user can compute the integral of any field in the dataset
+    along the given paths. The results of these path integrals are
+    stored in a cumulative fashion, which means that, once computed, the
+    accumulated value of the field is stored at each point along the
+    path, allowing the user to query this information.
+
+    Attributes
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:179]: I think you can get rid of these lines, since you don't fill them out, and generally we don't include stuff in the main docstring for these attributes anyway.

File: yt/visualization/accumulators.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+        self.paths      = []
+        self.ds         = ds
+        self.ad         = ds.all_data()
+        self.accum      = []
+        self.left_edge  = self.ds.domain_left_edge
+        self.right_edge = self.ds.domain_right_edge
+        # Make sure that the path has proper units. If no units are
+        # specified, assume code units
+        for p in paths:
+            if not isinstance(p, YTArray) or p.units == unyt.dimensionless:
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:201]:
```suggestion
            if not isinstance(p, YTArray) or p.units.is_dimensionless:
```

File: yt/visualization/accumulators.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+        -------
+        path_field_values : np.ndarray
+            The array containing the field values at each point along the
+            path.
+        """
+        node = tree.locate_node(path[idx])
+        # Data is a list, with one element for each field component
+        data = node.data.my_data
+        # Number of cells in the node
+        ncells = data[0].shape
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:258]: Maybe change the name of `ncells` to something like `dims` to indicate that it's not just number of cells, but cells-by-dimension.

File: yt/visualization/accumulators.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+            left_check = path[idx] < self.left_edge
+            right_check = path[idx] >= self.right_edge
+            if np.sum(np.logical_or(left_check, right_check)):
+                msg = f"Point `{path[idx]}` at index `{idx}` outside domain bounds."
+                msg += f"LE: `{self.left_edge}`, RE: `{self.right_edge}`"
+                raise ValueError(msg)
+            # Figure out which cell in the node the point falls within
+            # Origin of node can be offset from origin of origin of volume,
+            # so we have to subtract it off to get the right cell indices
+            cell_ind = ((path[idx] - node_left_edge) / cell_size).astype(int)
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:277]:
```suggestion
            cell_ind = ((path[idx] - node_left_edge) / cell_size).astype('i8')
```

File: yt/visualization/accumulators.py [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+            I = get_row_major_index(ncells, cell_ind)
+            # Get the value of each component for the current cell
+            vals[idx] = data[I] 
+            # See if next point is still within the same node (if the next point
+            # is still in range of the array)
+            idx += 1
+            if idx != npts:
+                left_check = path[idx] < node_left_edge
+                right_check = path[idx] >= node_right_edge
+                if np.sum(np.logical_or(left_check, right_check)):
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:290]:
```suggestion
                if np.any(left_check | right_check):
```

File: yt/visualization/accumulators.py (hunk 1 above)
Comment (jcoughlin11) [reply yt/visualization/accumulators.py:81]: Yeah, path is unitless. Looking at it now, this probably isn't a good way of doing it. I assume that the path is in code units. I think I need to handle the units better, as well as make path a unyt array. I'll do that.

File: yt/visualization/accumulators.py (hunk 3 above)
Comment (jcoughlin11) [reply yt/visualization/accumulators.py:122]: Not sure, as I've never heard of those numpy functions before haha! I'll take a look, and if they are the same then I'll use those! Thanks!

File: yt/visualization/accumulators.py (hunk 1 above)
Comment (jcoughlin11) [reply yt/visualization/accumulators.py:81]: No, I was wrong. path either has units specified by the user or, if no units are given by the user, then code length units are assumed. The above construction was done because np.dot was removing the units completely and I did not find a unyt implementation of np.dot. As such, I used .d to just access the quantities in the path and field arrays and then manually handled the units. There might be a better way to do that, though.

Comment (jcoughlin11) [reply yt/visualization/accumulators.py:81]: The units are manually added in on line 82

**Comment by matthewturk** [comment]
@jcoughlin11 Can you update answer store, and I'll pursue reviewing this in more detail?

**Comment by jcoughlin11** [comment]
@matthewturk Just did. I also merge with the pytest branch, so everything should be up to date

File: yt/visualization/accumulators.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+    ..math::
+
+        I = \int_C \phi(x1,x2,...,xn)d\vec{r}
+
+    Parameters
+    ----------
+    p : YTArray
+        The path to be integrated along
+
+    fieldVals : YTArray
```
Comment (cphyc) [inline yt/visualization/accumulators.py:101]:
```suggestion
    field_vals : YTArray
```

File: yt/visualization/accumulators.py [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+        data = ustack(data, axis=1)
+        # Cell width
+        node_left_edge = self.ds.arr(node.get_left_edge(), "code_length")
+        node_right_edge = self.ds.arr(node.get_right_edge(), "code_length")
+        cell_size = (node_right_edge - node_left_edge) / ncells
+        while idx < npts:
+            # Make sure point is within domain
+            left_check = path[idx] < self.left_edge
+            right_check = path[idx] >= self.right_edge
+            if np.sum(np.logical_or(left_check, right_check)):
```
Comment (cphyc) [inline yt/visualization/accumulators.py:256]:
```suggestion
            if np.any(left_check | right_check):
```

**Review by cphyc — COMMENTED** [review]
It would be great to run pre-commit.ci autofix on this one so that the formatting diff disappear.

**Comment by jcoughlin11** [comment]
@cphyc Sorry if this is a dumb question, but how do I run pre-commit.ci on this?

File: yt/visualization/accumulators.py (hunk 3 above)
Comment (jcoughlin11) [reply yt/visualization/accumulators.py:122]: It's the same as `np.ravel_multi_index` so I'll use that instead.

File: yt/visualization/accumulators.py (hunk 6 above)
Comment (jcoughlin11) [reply yt/visualization/accumulators.py:244]: Done.

File: yt/frontends/stream/tests/test_stream_unstructured.py [hunk 11]
```diff
@@ -8,31 +8,17 @@ def test_multi_mesh():
...
-        ],
-        dtype=np.int64,
-    )
-    connect2 = np.array(
-        [
-            [1, 2, 3],
-        ],
-        dtype=np.int64,
-    )
+    connect1 = np.array([[0, 1, 3]], dtype=np.int64)
```
Comment (matthewturk) [inline yt/frontends/stream/tests/test_stream_unstructured.py:11]: @jcoughlin11 hey, any idea why this got changed?  Did black get run differently?

File: yt/visualization/accumulators.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+    accum : YTArray
+        The cumulative value of the field integral at each path
+        segment
+    """
+    # https://en.wikipedia.org/wiki/Line_integral
+    # np.linalg.norm returns a ndarray, so when multiplied by field_vals, which
+    # is a YTArray, this leads to incorrect units. There does not appear to be
+    # a unyt implementation of norm, as far as I'm aware, so units will be
+    # handled manually for the time being
+    accum = np.cumsum(field_vals[:-1].d * np.linalg.norm(p[1:].d - p[:-1].d, axis=1))
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:115]: Why is it `field_vals[:-1].d`?  It kind of seems to me that if this is the case, it should all be offset by the initial value, shouldn't it?

File: yt/visualization/accumulators.py [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+    def __init__(self, paths, ds):
+        self.paths = []
+        self.ds = ds
+        self.ad = ds.all_data()
+        self.accum = []
+        self.left_edge = self.ds.domain_left_edge
+        self.right_edge = self.ds.domain_right_edge
+        # Make sure that the path has proper units. If no units are
+        # specified, assume code units
+        for p in paths:
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:142]: We need to make sure that paths, if it's a list of scalars, is converted to the right dimensionality.  I think the safest thing to do is to ensure (by validating the input) that `paths` is either a `YTArray` or a `np.ndarray`, and if it's a list, it needs to be a list of either scalars (in which case it gets dumped into a `(N,1)` array) or of ndarrays or ytarrays of all the same size.

I believe you can test this by checking if it's an array; if it is, then call `ensure_2d` on it.  If it's not -- if it's a list -- try converting it to a YTArray.

File: yt/visualization/accumulators.py [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+            If True, then `field` is a vector and it is assumed that it's
+            an iterable containing the x, y, and z components of the field.
+            Otherwise, the field is assumed to be a scalar field.
+
+        Returns
+        -------
+        tree : AMRKDTree
+            An AMRKDTree for the dataset containing the given field values.
+        """
+        tree = AMRKDTree(self.ds)
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:168]: Does this recreate the tree every time?  You could cache `tree` as an attribute or property, and just add fields as needed.

File: yt/visualization/accumulators.py [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+        use_log = [False for i in range(len(field))]
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:169]:
```suggestion
        use_log = [False for f in field]
```

But also, doesn't this mean that if `field` is a tuple of field type and field name, it will be two long?  How are you checking that field is a list?

File: yt/visualization/accumulators.py [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+        field = [self.ad._determine_fields(f)[0] for f in field]
+        tree.set_fields(field, use_log, False)
+        return tree
+
+    def _get_path_field_values(self, tree, path, idx, vals, npts):
+        r"""
+        Determines the value of the field at each point along the path.
+
+        Parameters
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:178]: these parameters are out of date -- what are `idx` `vals` `npts`?  I'm not sure I understand this.

File: yt/visualization/accumulators.py [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+        data = node.data.my_data
+        # Number of cells in the node
+        dims = data[0].shape
+        # Put the data in a more convenient form: dims x ndims array
+        data = [d.flatten() for d in data]
+        data = ustack(data, axis=1)
+        # Cell width
+        node_left_edge = self.ds.arr(node.get_left_edge(), "code_length")
+        node_right_edge = self.ds.arr(node.get_right_edge(), "code_length")
+        cell_size = (node_right_edge - node_left_edge) / dims
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:204]: So if it's *vertex* centered data, which I would encourage you to check, this will not be the cell size -- the cell size will be `/ (dims-1)` (that's not valid python code).

File: yt/visualization/accumulators.py [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+        while idx < npts:
+            # Make sure point is within domain
+            left_check = path[idx] < self.left_edge
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:207]: This could be applied outside of the loop.

File: yt/visualization/accumulators.py [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+            # Get the value of each component for the current cell
+            vals[idx] = data[I]
+            # See if next point is still within the same node (if the next point
+            # is still in range of the array)
+            idx += 1
+            if idx != npts:
+                left_check = path[idx] < node_left_edge
+                right_check = path[idx] >= node_right_edge
+                if np.any(left_check | right_check):
+                    vals, idx = self._get_path_field_values(tree, path, idx, vals, npts)
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:230]: It seems to me that this might also be doable outside of the loop.  So do a check if it's inside, and if it's not, call the function.

File: yt/visualization/accumulators.py [hunk 20]
```diff
@@ -0,0 +1,290 @@
...
+        # Number of cells in the node
+        dims = data[0].shape
+        # Put the data in a more convenient form: dims x ndims array
+        data = [d.flatten() for d in data]
+        data = ustack(data, axis=1)
+        # Cell width
+        node_left_edge = self.ds.arr(node.get_left_edge(), "code_length")
+        node_right_edge = self.ds.arr(node.get_right_edge(), "code_length")
+        cell_size = (node_right_edge - node_left_edge) / dims
+        while idx < npts:
```
Comment (matthewturk) [inline yt/visualization/accumulators.py:205]: I believe this entire loop can be vectorized.

I think what we should do is familiarize ourselves with the behavior -- make sure that we have passing and appropriate tests, that they cover some corner cases, etc, and then turn it into a vectorized loop.  But this definitely can be vectorized, because we're just selecting values here, not computing an integral (until later.)

**Review by matthewturk — CHANGES_REQUESTED** [review]
Looking much better.  I have some comments.

File: yt/visualization/accumulators.py (hunk 17 above)
Comment (jcoughlin11) [reply yt/visualization/accumulators.py:204]: How do I check that?

File: yt/visualization/accumulators.py (hunk 20 above)
Comment (jcoughlin11) [reply yt/visualization/accumulators.py:205]: The reason it's not vectorized is because it's recursive, since not every point will be in the same node, and I wasn't sure how to vectorize that.

File: yt/visualization/accumulators.py (hunk 18 above)
Comment (jcoughlin11) [reply yt/visualization/accumulators.py:207]: I'm not sure I understand how? `path[idx]` is a different point for each value of `idx`

File: yt/visualization/accumulators.py (hunk 19 above)
Comment (jcoughlin11) [reply yt/visualization/accumulators.py:230]: As above, I'm not sure I understand how, since the checks depend on the value of `idx`

**Comment by jcoughlin11** [comment]
@neutrinoceros Ok, thanks! I'm a bit confused, though, as black runs automatically as a part of precommit whenever I commit locally. If my precommit file is out of sync with that of main, is there no way around this sort of issue?

**Comment by matthewturk** [comment]
@jcoughlin11 you may want to try the rebase or whatever again

**Comment by neutrinoceros** [comment]
⚠️ rebasing with commits like "merge from main" on the branch is a very bad idea if you don't know what you're doing. This is why I advised against merging.

**Comment by neutrinoceros** [comment]
I admit I didn't check for that... @jcoughlin11 , can you give us a post-mortem here so we know what you just did ?

**Comment by jcoughlin11** [comment]
Whoops. ctrl+z!
